Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift

```diff
@@ -0,0 +1,434 @@
+        
+        // We should decide if we're can upgrade based on the first response header: if we aren't upgrading,
+        // by the time the body comes in we should be out of the pipeline. That means that if we don't think we're
+        // upgrading, the only thing we should see is a response head. Anything else in an error.
+        guard case .head(let response) = responsePart else {
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.invalidHTTPOrdering)
+            self.notUpgrading(context: context, data: responsePart)
+            return
+        }
+        
+        // Assess whether the upgrade response has accepted our upgrade request.
+        guard case .switchingProtocols = response.status else {
+            self.notUpgrading(context: context, data: responsePart)
+            return
+        }
+        
+        // Ok, we have a HTTP response. Check if it's an upgrade confirmation.
+        // If it's not, we want to pass it on and remove ourselves from the channel pipeline.
+        let acceptedProtocols = response.headers[canonicalForm: "upgrade"].map(String.init)
+        
+        guard acceptedProtocols.count > 0 else {
+            // There is no upgrade protocols returned dspite status being 'switching protocols'.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.responseProtocolNotFound)
+            self.notUpgrading(context: context, data: responsePart)
+            return
+        }
+
+        // This is an upgrade response.
+        // This may take a while, so while we're waiting more data can come in.
+        self.upgradeState = .awaitingUpgrader
+        
+        self.handleUpgrade(context: context, upgradeResponse: response, receivedProtocols: acceptedProtocols).whenSuccess { callback in
+
+            if let callback = callback {
+                self.gotUpgrader(upgrader: callback)
+            } else {
+                self.notUpgrading(context: context, data: responsePart)
+            }
+        }
+    }
+    
+    private func handleUpgrade(context: ChannelHandlerContext, upgradeResponse response: HTTPResponseHead, receivedProtocols: [String]) -> EventLoopFuture<(() -> Void)?> {
+        
+        let allHeaderNames = Set(response.headers.map { $0.name.lowercased() })
+        
+        // We now set off a chain of Futures to try to find a protocol upgrade.
+        // While this is blocking, we need to buffer inbound data.
+        let protocolIterator = receivedProtocols.makeIterator()
+        
+        return self.handleUpgradeForProtocol(context: context, protocolIterator: protocolIterator, response: response, allHeaderNames: allHeaderNames)
+    }
+    
+    /// Attempt to upgrade a single protocol.
+    private func handleUpgradeForProtocol(context: ChannelHandlerContext, protocolIterator: Array<String>.Iterator, response: HTTPResponseHead, allHeaderNames: Set<String>) -> EventLoopFuture<(() -> Void)?> {
```

Comment:
This function only makes synchronous function calls, and only returns pre-succeeded futures. That means you can remove the `Future` wrapping from it entirely and just have it be a synchronous function. That will help avoid some confusion in some other functions.